### PR TITLE
Add cardinality logic, spot price and mutable fees

### PR DESCRIPTION
**Pull request by jacopo-eth**
This PR introduces multiple updates over core logic of the price feed.

## Cardinality
- Added `lastUpdatedCardinality` in `PoolData` struct and `MAX_CARDINALITY`
- Added logic to increase observationCardinalityNext when updating pools, according to `cardinalityNextIncrease` param specified by the caller

## Spot price
- Added possibility to set `secondsTwapInterval` to 0, calculating quotes from the spot price of a pool

## Fees
- Introduced `addFee` to permissionlessly introduce new fee tiers when supported by Uniswap, and use them during `updatePool`

## Other
- Refactored code and renamed variables
- Updated `IPriceFeed` interface
- Updated readme

---

### Notes
- `MAX_CARDINALITY` value has been hardcoded to 256.

### CODE REVIEW

**Comment by jjranalli** [comment]
### Slice distribution request

- jacopo.eth : 5000

File: contracts/PriceFeed.sol [hunk 1]
```diff
@@ -75,147 +82,239 @@ contract PriceFeed is IPriceFeed {
...
-        arithmeticMeanTick,
-        baseAmount,
-        baseToken,
-        quoteToken
-      );
+      // Get spot price
+      if (secondsTwapInterval == 0) {
+        // Get sqrtPriceX96 from slot0
+        (uint160 sqrtPriceX96, , , , , , ) = IUniswapV3Pool(pool).slot0();
+        quoteAmount = _getQuoteAtSqrtPriceX96(sqrtPriceX96, baseAmount, baseToken, quoteToken);
```
Comment (drgorillamd) [inline contracts/PriceFeed.sol:108]: This returns the *price* (and the amount you'd for a given tick), not a real *quote* -> this price is only valid within a tick, if the amountIn is enough, price might change (ie liquidity from the tick is consumed and the next non-empty tick is used) resulting in a smaller amount out

-> To get a real quote, there is no other choice than simulating an actual swap (ie try-catch calling a contract swapping and reverting while returning the amount received) -> quite costly if onchain tho:/

Comment (jjranalli) [reply contracts/PriceFeed.sol:108]: If with _quote_ you mean the amount that you'd get by swapping, yes that's how it should be. The price feed is designed to return the absolute amount for a currency pair, and should be unrelated to swaps.

You can also find this in the gotchas FAQs
> Quotes represent a time-weighted average price for a currency in a certain amount of time (see [TWAP oracles](https://docs.uniswap.org/protocol/concepts/V3-overview/oracle)). As such they don't exactly correspond to the amount displayed during a swap on Uniswap, but they're better suited to estimate the value of a currency.

Comment (drgorillamd) [reply contracts/PriceFeed.sol:108]: @jjranalli I meant the spot price (used to get a quote), not the twap;)

Comment (drgorillamd) [reply contracts/PriceFeed.sol:108]: This is a correct spot price but cannot be used to get a quote (ie an amount out), as crossing ticks changes it

Comment (jjranalli) [reply contracts/PriceFeed.sol:108]: > I meant the spot price

true. Should modify the gotcha to be more generic and explicitly refer to spot price too.

> spot price used to get the quote, which is "how many token out do I get for amount in?"

Thing is, the price feed quote is "how many `quoteToken`s is a `baseAmount` of `baseToken`s worth". This is independent of liquidity / inAmount, and doesn't aim to estimate a potential swap. Does it make sense?

Comment (jjranalli) [reply contracts/PriceFeed.sol:108]: closing readme gotcha has been updated

**Pull request merged**